### ISSUE TITLE
Performance

### ISSUE THREAD

**Issue by simnalamburt** [simnalamburt/vim-mundo#11]
For some reason, `vim-mundo` just got very slow for large files. This should be fixed.

**Comment by dsummersl** [comment]
Definitely - any good example file/size?

**Comment by simnalamburt** [comment]
I was working with some quite large file like 100KB and It took me more than 5 second to open Gundo tab. I'm going to do some profiling if this problem occurs again to me.

**Comment by starcraftman** [comment]
@simnalamburt I just noticed the sluggishness too. It has to do with the inline_undo functionality. If this can't be made faster, it should probably be opt in.

**Comment by dsummersl** [comment]
Oh funny I thought it was. I'll take a look. Thx for the feedback.

**Comment by starcraftman** [comment]
@dsummersl No problem. Thanks for picking up where sjl left off. Odd how he just stopped, just like kien. :/

**Comment by dsummersl** [comment]
I'm going to leave this open - clearly that feature needs to be improved.

**Comment by Gordin** [comment]
I'm not sure if https://github.com/simnalamburt/vim-mundo/commit/47d95c19b6be69260a1d448da1c54a9d3351955c was supposed to fix this completely, but my vim is still completely unusable while the Gundo tab is open in the current version. Not only in the Gundo window, also everything else is slow, just doing j/k movements in any other window takes about 3 seconds before the cursor moves. Also, I'm not sure if this is the same bug because it doesn't seem to be connected to large file size, but directly correlated to the size of the history. I've tested this on different files, and it seems like on my machine everything you do takes about 1 second longer for every 1000 histery entries while Gundo is open.

**Comment by starcraftman** [comment]
@Gordin Could be perhaps I shouldn't have been so definite saying what the origin was. I was only indicating what I thought was making this fork slower then gundo. Gundo itself has always been a bit sluggish going j/k imo.

Ultimately, probably only gonna get fixed until somebody does profiling/tracing of the func calls.

**Comment by dsummersl** [comment]
@Gordin definitely believe that - I know that gundo/mundo both draw the entire history. I'm not sure how amenable the graphlog algorithm is to partial redrawing, but I like the idea of only computing/drawing that which is on screen. There are more memoization opportunities, for sure.  I'll look into this as soon as I can, but if you see something or have suggestions I'm all ears :)

**Comment by Gordin** [comment]
@dsummersl I didn't look into the code at all but this sounds like Gundo is recomputing the tree every time anything happens in vim? If this is the case, thats the problem, isn't it? I'm not sure how showing the graph is handled, but it seems like most problems would be fixed by just not doing anything unless an item is added to the untotree or the Gundo tab needs to scroll?

**Comment by dsummersl** [comment]
https://github.com/simnalamburt/vim-mundo/pull/13

@Gordin I revisited the source and discovered that there were already guards to do what you describe, but they weren't working. I've made a PR that fixes this.

WRT repainting only whats visible I guess what I was trying to describe would address startup time. If the graph algorithm supports partial tree rendering (maybe it does?), then we could incrementally generate the graph minimally to show what is visible on the screen...that would make startup time the same for undo histories of 1000 or 100 (I definitely still see a lag if I open mundo with a large undo history - navigating is definitely better with this PR tho). I'll open a separate issue for this...

If anybody has time to try out this PR I'd appreciate it...otherwise I'll play with it a couple more days before I merge it in.
